dev/src/bundle.ts

```diff
@@ -0,0 +1,184 @@
+  // Resulting documents for the bundle, keyed by full document path.
+  private documents: Map<string, BundledDocument> = new Map();
+  // Named queries saved in the bundle, keyed by query name.
+  private namedQueries: Map<string, firestore.INamedQuery> = new Map();
+
+  // The latest read time among all bundled documents and queries.
+  private latestReadTime = new Timestamp(0, 0);
+
+  constructor(private bundleId: string) {}
+
+  add(documentSnapshot: DocumentSnapshot): BundleBuilder;
+  add(queryName: string, querySnap: QuerySnapshot): BundleBuilder;
+  /**
+   * Adds a Firestore document snapshot or query snapshot to the bundle.
+   * Both the documents data and the query read time will be included in the bundle.
+   *
+   * @param {DocumentSnapshot | string=} documentOrName A document snapshot to add or a name of a query.
+   * @param {Query?=} querySnapshot A query snapshot to add to the bundle, if provided.
+   * @returns {BundleBuilder} This instance.
+   *
+   * @example
+   * let bundle = firestore.bundle('data-bundle');
+   * const bundleStream =
+   *    await bundle.add(await firestore.doc('abc/123').get()) // Add a document
+   *                .add('coll-query', await firestore.collection('coll').get()) // Add a named query.
+   *                .stream();
+   * // Stream `bundleStream` to clients.
+   */
+  add(
+    documentOrName: DocumentSnapshot | string,
+    querySnapshot?: QuerySnapshot
+  ): BundleBuilder {
+    if (documentOrName instanceof DocumentSnapshot) {
+      this.addBundledDocument(documentOrName);
+    } else if (querySnapshot instanceof QuerySnapshot) {
+      this.addNamedQuery(documentOrName, querySnapshot);
+    }
+
+    return this;
+  }
+
+  private addBundledDocument(snap: DocumentSnapshot) {
+    if (snap.exists) {
+      const docProto = snap.toDocumentProto();
+      if (
+        !this.documents.has(snap.id) ||
+        Timestamp.fromProto(this.documents.get(snap.id)!.metadata.readTime!) <
+          snap.readTime
+      ) {
+        this.documents.set(snap.id, {
+          document: docProto,
+          metadata: {
+            documentKey: docProto.name,
+            readTime: snap.readTime.toProto().timestampValue,
+          },
+        });
+      }
+      if (snap.readTime > this.latestReadTime) {
+        this.latestReadTime = snap.readTime;
+      }
+    }
+  }
+
+  private addNamedQuery(name: string, querySnap: QuerySnapshot) {
+    if (this.namedQueries.has(name)) {
+      throw new Error(`Query name conflict: ${name} is already added.`);
+    }
+
+    this.namedQueries.set(name, {
+      name,
+      bundledQuery: querySnap.query.toBundledQuery(),
+      readTime: querySnap.readTime.toProto().timestampValue,
+    });
+
+    for (const snap of querySnap.docs) {
+      this.addBundledDocument(snap);
+    }
+
+    if (querySnap.readTime > this.latestReadTime) {
+      this.latestReadTime = querySnap.readTime;
+    }
+  }
+
+  private lengthPrefixedBuffer(payload: string): Buffer {
```

Comment:
Please add some JSDoc here, as the method is not 100% self explanatory.

This is also missing a return type.